File: C++/CANifier Demo/src/Tasks/TaskPWMmotorController.cpp
#include <Tasks/TaskPWMmotorController.h>
#include "ctre/phoenix/LinearInterpolation.h"
#include "Platform/Platform.h"

TaskPWMmotorController::~TaskPWMmotorController() {}
/* ILoopable */
void TaskPWMmotorController::OnStart() {
	/* If already running, do nothing */
	if(_running)
		return;

	/* Start transmitting neutral signal */
    _percentOutput = 0;
    Hardware::canifier->SetPWMOutput(Constants::MotorControllerCh, 0);
    Hardware::canifier->EnablePWMOutput(Constants::MotorControllerCh, true);

    /* Task is now running */
    _running = true;
}
void TaskPWMmotorController::OnStop() {
	/* Stop transmitting PWM */
	Hardware::canifier->EnablePWMOutput(Constants::MotorControllerCh, false);

	/* Task has stopped, now take note */
	_running = false;
}
bool TaskPWMmotorController::IsDone() {
	return false;
}
void TaskPWMmotorController::OnLoop(){
    /* just grab three axis and direct control the components */
    float axis = Hardware::gamepad->GetRawAxis(Constants::Gamepad_y);
    /* scale to typical PWM widths */
    float pulseUs = CTRE::LinearInterpolation::Calculate(axis, -1, 1000, +1, 2000); /* [-1,+1] => [1000,2000]us */
    /* scale to period */
    float periodUs = 4200; // hard-coded for now, this will be settable in future firmware update.
    _percentOutput = pulseUs / periodUs;
    /* set it */
    Hardware::canifier->SetPWMOutput(Constants::MotorControllerCh, _percentOutput);
}

File: Java/CANifier Demo/src/main/java/org/usfirst/frc/team3539/robot/Tasks/TaskPWMmotorController.java
package org.usfirst.frc.team3539.robot.Tasks;

import com.ctre.phoenix.ILoopable;
import org.usfirst.frc.team3539.robot.Framework.LinearInterpolation;
import org.usfirst.frc.team3539.robot.Platform.Constants;
import org.usfirst.frc.team3539.robot.Platform.Hardware;

public class TaskPWMmotorController implements ILoopable {
	float _percentOut;
	boolean _running; /* Assists TaskMainLoop with tracking, used within init */

	/* ILoopable */
	public void onStart() {
		/* If we are already running, nothing to do */
		if (_running) {
			return;
		}

		/* Start transmitting neutral */
		_percentOut = 0;
		Hardware.canifier.setPWMOutput(Constants.kMotorControllerCh.value, 0);
		Hardware.canifier.enablePWMOutput(Constants.kMotorControllerCh.value,
				true);

		/* Task is now running */
		_running = true;
	}

	public void onStop() {
		/* Stop transmitting PWM */
		Hardware.canifier.enablePWMOutput(Constants.kMotorControllerCh.value,
				false);

		/* Task has stopped */
		_running = false;
	}

	public boolean isDone() {
		return false;
	}

	public void onLoop() {
		/* Grab three axis and direct control the PWM MotorController */
		float axis = (float) Hardware.gamepad
				.getRawAxis(Constants.GamePadAxis_y);
		/* Scale to typical PWM widths */
		float pulseUs = LinearInterpolation.calculate(axis, -1, 1000f, +1,
				2000f); /* [-1,+1] => [1000,2000]us */
		/* Scale to period */
		float periodUs = 4200; /*
								 * Hard-coded for now, this will be settable in
								 * future firmware update.
								 */
		_percentOut = pulseUs / periodUs;
		/* Set PWM Motor Controller */
		Hardware.canifier.setPWMOutput(Constants.kMotorControllerCh.value,
				_percentOut);
	}

	public String toString() {
		return "TaskPWMmotorController:";
	}
}
